src/lib/shell/MainLoopZephyr.cpp

```diff
@@ -14,30 +14,117 @@
  *    See the License for the specific language governing permissions and
  *    limitations under the License.
  */
+#include <cstring>
 #include <init.h>
 #include <shell/shell.h>
+#include <stdio.h>
 
 #include <lib/core/CHIPError.h>
 #include <lib/shell/Engine.h>
 #include <lib/shell/streamer_zephyr.h>
+#include <lib/support/CodeUtils.h>
 
+using chip::Shell::cmd_completion_context;
 using chip::Shell::Engine;
+using chip::Shell::shell_command_t;
+using chip::Shell::shell_map_t;
 
 static int cmd_matter(const struct shell * shell, size_t argc, char ** argv)
 {
     chip::Shell::streamer_set_shell(shell);
     return (Engine::Root().ExecCommand(argc - 1, argv + 1) == CHIP_NO_ERROR) ? 0 : -ENOEXEC;
 }
 
+/**
+ * This recursive function obtains all sub-commands (possbile completions) for a command node
+ * and build it into a tree of Zephyr's shell_static_entry structure.
+ *
+ * @param prefix                  The path/prefix until the root command node of the tree, without tailing space.
+ *                                e.g. "dns browse". Set to emptry string "" if building tree for a root-level command.
+ * @param syntax                  The single-word command name/syntax for the root node of the tree. e.g. "resolve".
+ *                                Set to emptry string "" if calling to build from the root (to include all commands).
+ * @param help                    The help text for the command at roote node of the tree. Set to NULL if none.
+ *
+ * @return                        The shell_static_entry structure tree that chains all the sub-commands from the
+ *                                requested command node.
+ *
+ */
+const struct shell_static_entry * build_command_tree(const char * prefix, const char * syntax, const char * help)
+{
+
+    char * sub_prefix = new char[strlen(prefix) + strlen(syntax) + 2];
```

Comment:
use ScopedMemoryBuffer here. Otherwise tracking if sub_prefix is allocated and freed correctly is hard.
Also add a guard to see if memory allocation succeeded.